kda/diagrams.py

```diff
@@ -459,69 +481,39 @@ def generate_flux_diagrams(G, cycle):
-    # get all of the unique edges in the input diagram
-    G_edges = _find_unique_edges(G)
+    cycle_edges = _append_reverse_edges(cycle_edges)
     # get edges that are uncommon between cycle and G
-    non_cycle_edges = _find_unique_uncommon_edges(G_edges, cycle)
+    non_cycle_edges = _find_unique_uncommon_edges(G, cycle_edges)
     # number of non-cycle edges in flux diagram
-    n_non_cycle_edges = G.number_of_nodes() - len(cycle_edges)
+    n_non_cycle_edges = G.number_of_nodes() - len(cycle)
     # generate all combinations of valid edges
     # TODO: generates too many edge lists: some create cycles, some
     # use both forward and reverse edges
     flux_diagrams = []
     for edge_list in itertools.combinations(non_cycle_edges, r=n_non_cycle_edges):
         # convert each edge list into numpy array
         edge_list = np.asarray(edge_list, dtype=int)
-        # initialize empty list for storing uni-directional edges
+        # collect the directional edges
         dir_edges = []
         for target in cycle:
-            # for each node in the cycle, collect the directional
-            # connection dictionary
-            cons = _get_directional_connections(target, edge_list)
-            if cons:
-                # if there are directional connections, generate and
-                # store the directional edges
-                dir_edges.append(_get_directional_edges(cons))
-        # flatten the nested lists of edges
-        dir_edges = [edge for edges in dir_edges for edge in edges]
+            path_edges = _get_flux_path_edges(target, edge_list)
+            dir_edges.extend(path_edges)
         if _flux_edge_conditions(dir_edges, n_non_cycle_edges):
-            # make a copy of the base graph
-            flux_diag = base_graph.copy()
-            # add all directional edges to flux diagram
+            # initialize a graph object
+            flux_diag = nx.MultiDiGraph()
+            # add cycle edges to list of directional edges
+            dir_edges.extend(cycle_edges)
+            # add all edges to flux diagram
             flux_diag.add_edges_from(dir_edges)
-            # collect all nodes in the potential flux
-            # diagram from the diagram edges
-            included_nodes = np.unique(flux_diag.edges)
-            # if the diagram contains all nodes it is still valid
-            if included_nodes.size == G.number_of_nodes():
-                # count how many cycles are in the flux diagram by removing
-                # 2-node cycles
-                # NetworkX stores forward/reverse cycles, so if 2 remain there
-                # is 1 unique cycle for this flux diagram
-                contains_1_cycle = (
-                    len([c for c in nx.simple_cycles(flux_diag) if len(c) > 2]) == 2
-                )
-                # if there is exactly 1 unique cycle in the
-                # generated diagram, it is valid
-                if contains_1_cycle:
```

Comment:
Since we are removing these checks in the `generate_flux_diagrams()` function itself, we should at least add them in the test that iterates over difficult flux diagram cases as a regression guard. If we later make a change to this algorithm (likely) it would be helpful to know that these cycles and nodes are correct.

Reply:
This was addressed in the latest commit.